core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java

```diff
@@ -1634,71 +1674,221 @@ boolean resumeSyncForDocument(
-    return UpdateResult.acknowledged(1, 1L, null);
+    return UpdateResult.acknowledged(1, 1L, updateOptions.isUpsert() ? documentId : null);
+  }
+
+  /**
+   * Update all documents in the collection according to the specified arguments.
+   *
+   * @param filter a document describing the query filter, which may not be null.
+   * @param update a document describing the update, which may not be null. The update to
+   *               apply must include only update operators.
+   * @return the result of the update many operation
+   */
+  UpdateResult updateMany(final MongoNamespace namespace,
+                          final Bson filter,
+                          final Bson update) {
+    return updateMany(namespace, filter, update, new UpdateOptions());
+  }
+
+  /**
+   * Update all documents in the collection according to the specified arguments.
+   *
+   * @param filter        a document describing the query filter, which may not be null.
+   * @param update        a document describing the update, which may not be null. The update to
+   *                      apply must include only update operators.
+   * @param updateOptions the options to apply to the update operation
+   * @return the result of the update many operation
+   */
+  UpdateResult updateMany(
+      final MongoNamespace namespace,
+      final Bson filter,
+      final Bson update,
+      final UpdateOptions updateOptions) {
+    // fetch all of the documents that this filter will match
+    final Map<BsonValue, BsonDocument> idToBeforeDocumentMap = new HashMap<>();
+    final BsonArray ids = new BsonArray();
+    this.getLocalCollection(namespace)
+        .find(filter)
+        .forEach(new Block<BsonDocument>() {
+          @Override
+          public void apply(@NonNull final BsonDocument bsonDocument) {
+            final BsonValue documentId = BsonUtils.getDocumentId(bsonDocument);
+            ids.add(documentId);
+            idToBeforeDocumentMap.put(documentId, bsonDocument);
+          }
+        });
+
+    // do the bulk write
+    final UpdateResult result = this.getLocalCollection(namespace)
+        .updateMany(filter, update, updateOptions);
+
+    // if this was an upsert, create the post-update filter using
+    // the upserted id.
+    // else, use the matched ids from prior to create a new filter
+    final BsonDocument updatedFilter;
+    if (result.getUpsertedId() != null) {
+      updatedFilter = getDocumentIdFilter(result.getUpsertedId());
+    } else {
+      updatedFilter = new BsonDocument("_id", new BsonDocument("$in", ids));
+    }
+
+    this.getLocalCollection(namespace).find(updatedFilter).forEach(new Block<BsonDocument>() {
+      @Override
+      public void apply(@NonNull final BsonDocument afterDocument) {
+        final BsonDocument beforeDocument;
+        final BsonValue documentId = BsonUtils.getDocumentId(afterDocument);
+
+        if ((beforeDocument = idToBeforeDocumentMap.get(documentId)) == null
```

Comment:
[nit] Assigning to beforeDocument like this feels kind of odd as opposed to just doing it right before the if

Reply:
👍